README.md
# Migrated to https://codeberg.org/1-1sam/lockatty

# lockatty
<a> <img src=lockatty-scrot.png></a>

lockatty is a simple screenlocker for Linux terminals.
## Usage
By default, lockatty will use your user's system password for authentication. If you wish to use a seperate password
just for lockatty, you can use the -l option which will prompt you to set a password and save its hash to ~/.pw. If
you wish to reset this password, you can delete the .pw file and it will prompt you again when the -l option is
invoked again.

The -e option can be used to print the contents of a given ascii file to the prompt like in the example screenshot. The
text prompt asking for your password can be disabled with the -n option. The text informing you of an incorrect password
attempt will still be printed, however.
## Install
Required dependencies are openssl and ncurses.

This will install it into /usr/local.
```bash
make
make install # as root
make clean
```
You may edit the DESTDIR and PREFIX variable to install it in your preferred directory.
```bash
make install DESTDIR=build PREFIX=/usr
```
It is recommended you install the documentation.
```bash
make install-doc DESTDIR=build PREFIX=/usr
```

If you wish to uninstall:
```bash
make uninstall # Make sure you specify DESTDIR and PREFIX if they were used during the make install.
```

## TODO

* [X] Add support for options.
* [X] Add support for user's system passwords.
* [ ] Add support for other hashing algorithims besides md5.
* [ ] Package for the AUR and SlackBuilds.org.

## License
[GPL-3](https://www.gnu.org/licenses/gpl-3.0-standalone.html)


CHANGELOG.md
# Changelog -- lockatty 1.0.0

## Added
* Added program options: -e or --echo, -l or --lockatty-pw, -n or --no-text-prompt, -? or --help, --usage, and
-V or --version.
* Created a manual.
* Support for users' system passwords.
* Warning for passwords that exceed the 50 character limit.

## Changes
* Separated lockatty.c into multiple c and header files in src directory.
* The echo file is now a program option which accepts any ascii file passed to it.
* ~/.lockatty/pw is now ~/.pw.
* Updated Makefile.
* Updated README.md.

## Removed
* Removed ~/.lockatty folder.
* Removed ~/.lockatty/echo file. Please use the -e option from now on.

## Fixes
* Fixed cat_echo(). Can now properly read files over 1024 bytes. Hopefully, at least.



src/pw.c
/*
Various password handling functions used by lockatty.
Copyright 2023, Samuel Young

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.

Contact: samyoung12788@gmail.com
*/

#include <stdio.h>
#include <ncursesw/ncurses.h>
#include <openssl/md5.h>
#include <string.h>

#include "file.h"

#define STRCMP_EQUAL 0
#define MAX_PW_LEN 50
#define MD5_LEN 32

void create_pw_file(char* str) {

    FILE* password_file;
    unsigned char digest[16];

    password_file = fopen(".pw","w");

    MD5(str, strlen(str), digest);

    /* Send the digest to .lockatty/pw in a hexadecimal format. */
    for (int i = 0; i < 16; i++) {
        fprintf(password_file, "%02x", (unsigned int) digest[i]);
    }

    fclose(password_file);
}

int lpwcmp(char* guess) {
    FILE* password_file;
    char buf[MD5_LEN + 1];
    char guess_md5[MD5_LEN + 1];
    unsigned char guess_digest[16];
    char tmp02x[6];

    password_file = fopen(".pw","r");

    fread(buf,MD5_LEN,1,password_file);
    buf[MD5_LEN] = '\0';

    MD5(guess, strlen(guess), guess_digest);

    /* Setting the first character in character_md5 to 0 will prevent a memory leak. */
    guess_md5[0] = '\0';

    /* Writes the hexadecimal number of guess_digest[i] to tmp02x, which is then appended to guess_md5 */
    for (int i = 0; i < 16; i++) {
        sprintf(tmp02x,"%02x",(unsigned int) guess_digest[i]);
        strcat(guess_md5,tmp02x);
    }

    guess_md5[MD5_LEN] = '\0';

    if ((strcmp(buf,guess_md5)) == STRCMP_EQUAL) {
        return 1;
    }
    return 0;
}


src/file.c
/*
Various file handling functions for lockatty.
Copyright 2023, Samuel Young

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.

Contact: samyoung12788@gmail.com
*/

#include <stdio.h>
#include <string.h>
#include <ncursesw/ncurses.h>

int check_file(char* filename) {
	FILE* file;
	if((file = fopen(filename, "r"))) {
		fclose(file);
		return 1;
	}
	return 0;
}

void cat_echo(char* file) {
	FILE* echo;
	char buf[1024];

	echo = fopen(file,"r");

	do {
		memset(buf, 0, 1023);
		fread(buf, 1, 1023, echo);
		printw("%s",buf);
	} while (buf[1022]);
	printw("\n");
	fclose(echo);
}


src/main.c
/*
lockatty, a simple terminal screenlocker.
Copyright 2023, Samuel Young

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.

Contact: samyoung12788@gmail.com
*/

#include <ncursesw/ncurses.h>
#include <stdlib.h>
#include <openssl/md5.h>
#include <string.h>
#include <unistd.h>
#include <sys/stat.h>
#include <argp.h>

#include "pw.h"
#include "file.h"
#include "shadow-auth.h"

#define STRCMP_EQUAL 0
#define MAX_PW_LEN 50
#define MD5_LEN 32
#define FOREVER_MAN 1

/* Getting argp ready */
const char* argp_program_version = "lockatty 1.0.0";
const char* argp_program_bug_address = "samyoung12788@gmail.com";
static char doc[] = "lockatty -- Simple terminal screen locker";
static char args_doc[] = "";

static struct argp_option options[] = {
	{ "lockatty-pw", 'l', 0, 0, "Use a lockatty-specific password rather than a user's system password." },
	{ "echo", 'e', "FILE", 0, "Echo the contents of an ascii file for the password prompt. Must pass absolute path for file." },
	{ "no-text-prompt", 'n', 0, 0, "Does not display the text prompt asking for the password."},
	{ 0 }
};

struct arguments {
	enum { SYSTEM_PW, LOCKATTY_PW } pw_type;
	int echoing;
	char* echo_file;
	int text_prompt;
};

static error_t parse_opt (int key, char* arg, struct argp_state* state) {
	struct arguments* arguments = state->input;

	switch (key) {
	case 'l':
		arguments->pw_type = LOCKATTY_PW; break;
	case 'e':
		arguments->echoing = 1;
		arguments->echo_file = arg; break;
	case 'n':
		arguments->text_prompt = 0; break;
	case ARGP_KEY_ARG:
		return 0; break;
	default:
		return ARGP_ERR_UNKNOWN; break;
	}
	return 0;
}

static struct argp argp = { options, parse_opt, args_doc, doc };
/* Done setting argp up. */

int main(int argc, char** argv) {

    char pwguess[MAX_PW_LEN];
	char pw1[MAX_PW_LEN];
	char pw2[MAX_PW_LEN];
	int auth;

	struct arguments arguments;
	arguments.pw_type = SYSTEM_PW;
    arguments.echoing = 0;
    arguments.text_prompt = 1;

    argp_parse(&argp, argc, argv, 0, 0, &arguments);

    /* Flying to the home directory for the winter. */
    chdir(getenv("HOME"));

    initscr();
    noecho();

	/* Create the pw file if it does not exist already. Only if using a lockatty password. */
	if (arguments.pw_type == LOCKATTY_PW) {
		printw("A lockatty password has not been set yet. Please enter the password you would like to you.\n");
		while (check_file(".pw") == 0) {
			getstr(pw1);
			printw("Please enter your password again.\n");
			getstr(pw2);
			if ((strcmp(pw1,pw2)) == 0 && strlen(pw1) <= 51) {
				create_pw_file(pw1);
				printw("Password set.\n");
			} else {
				printw("Passwords do not match or are over 50 characters long.\n");
			}
		}
	}

	if (arguments.echoing && check_file(arguments.echo_file)) {
		cat_echo(arguments.echo_file);
	}

	if (arguments.text_prompt) {
		printw("Please enter your password.\n");
		refresh();
	}

	for (;;) {
		getstr(pwguess);

		if (arguments.pw_type == SYSTEM_PW) {
			auth = shadow_auth(pwguess);
		} else if (arguments.pw_type == LOCKATTY_PW) {
			auth = lpwcmp(pwguess);
		}

		if (auth) {
			endwin();
			return 0;
		} else {
			printw("Incorrect.\n");
		}
	}
	return 0;
}


src/shadow-auth.c
/*
Shadow authentication function used by lockatty.
Copyright 2023, Samuel Young

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.

Contact: samyoung12788@gmail.com
*/

#define _XOPEN_SOURCE

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <crypt.h>
#include <shadow.h>

int shadow_auth(char* pwg) {
	char* crypt_pwg;
	struct spwd* spw;
	char* user = getenv("USER");

	spw = getspnam(user);

	crypt_pwg = crypt(pwg, spw->sp_pwdp);

	if (strcmp(crypt_pwg,spw->sp_pwdp) == 0) {
		return 1;
	} else {
		return 0;
	}
}


lockatty.c
/*
lockatty, a simple terminal screenlocker.
Copyright 2023, Samuel Young

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.

Contact: samyoung12788@gmail.com
*/

#include <ncursesw/ncurses.h>
#include <stdlib.h>
#include <openssl/md5.h>
#include <string.h>
#include <unistd.h>
#include <sys/stat.h>

#define STRCMP_EQUAL 0
#define MAX_PW_LEN 50
#define MD5_LEN 32
#define FOREVER_MAN 1

void check_pw_file();
void create_pw_file(char* str);
int check_file(char* filename);
void cat_echo();
int pwcmp(char* guess);

int main(void) {

    char pwguess[MAX_PW_LEN];

    /* Flying to the home directory for the winter. */
    chdir(getenv("HOME"));

    /* Create .lockatty in home if it does not exist. This is where all user files are stored. */
    if (check_file(".lockatty/") == 0) {
        mkdir(".lockatty",0755);
    }

    initscr();
    noecho();

    check_pw_file();

    if (check_file(".lockatty/echo")) {
        cat_echo();
    }

    while (FOREVER_MAN) {
        printw("Please enter your password:\n");
        refresh();
        getstr(pwguess);

        if ((pwcmp(pwguess))) {
            /* Correct password */
            endwin();
            return 0;
        } else {
            printw("Incorrect. Try again. \n");
            refresh();
        }
    }
    return 1;
}

void check_pw_file() {
    char pw1[MAX_PW_LEN];
    char pw2[MAX_PW_LEN];
    while (check_file(".lockatty/pw") == 0) {

        printw("A password has not been set yet. Please enter the password you would like to use:\n");
        getstr(pw1);
        printw("Enter your password again.\n");
        getstr(pw2);
        if ((strcmp(pw1,pw2)) == STRCMP_EQUAL) {
            create_pw_file(pw1);
            printw("Password set.\n");
        } else {
        	printw("Passwords do match.\n");
        }
    }
}

void create_pw_file(char* str) {

    FILE* password_file;
    unsigned char digest[16];

    password_file = fopen(".lockatty/pw","w");

    MD5(str, strlen(str), digest);

    /* Send the digest to .lockatty/pw in a hexadecimal format. */
    for (int i = 0; i < 16; i++) {
        fprintf(password_file, "%02x", (unsigned int) digest[i]);
    }

    fclose(password_file);
}

int check_file(char* filename) {
    FILE *file;
    if((file = fopen(filename, "r"))) {
        fclose(file);
        return 1;
    }
    return 0;
}

void cat_echo() {
    FILE* echo;
    char buf[1024];

    fopen(".lockatty/echo","r");

    buf[1023] = '\0'; /* This terminating null should never be overwritten. */
    do {
    	memset(buf, 0, 1022);
        fread(buf, 1, 1023, echo);
        printw("%s",buf);
    } while (buf[1022]);
	printw("\n");
    fclose(echo);
}

int pwcmp(char* guess) {
    FILE* password_file;
    char buf[MD5_LEN + 1];
    char guess_md5[MD5_LEN + 1];
    unsigned char guess_digest[16];
    char tmp02x[6];

    password_file = fopen(".lockatty/pw","r");

    fread(buf,MD5_LEN,1,password_file);
    buf[MD5_LEN] = '\0';

    MD5(guess, strlen(guess), guess_digest);

    /* Setting the first character in character_md5 to 0 will prevent a memory leak. */
    guess_md5[0] = '\0';

    /* Writes the hexadecimal number of guess_digest[i] to tmp02x, which is then appended to guess_md5 */
    for (int i = 0; i < 16; i++) {
        sprintf(tmp02x,"%02x",(unsigned int) guess_digest[i]);
        strcat(guess_md5,tmp02x);
    }

    guess_md5[MD5_LEN] = '\0';

    if ((strcmp(buf,guess_md5)) == STRCMP_EQUAL) {
        return 1;
    }
    return 0;
}
